--- newlib/newlib/libc/stdlib/mprec.c
+++ newlib/newlib/libc/stdlib/mprec.c
@@ -93,6 +93,9 @@
 #define _Kmax 15
 */
 
+void balloc_lock();
+void balloc_unlock();
+
 _Bigint *
 _DEFUN (Balloc, (ptr, k), struct _reent *ptr _AND int k)
 {
@@ -102,12 +105,14 @@
   _REENT_CHECK_MP(ptr);
   if (_REENT_MP_FREELIST(ptr) == NULL)
     {
+      balloc_lock();
       /* Allocate a list of pointers to the mprec objects */
       _REENT_MP_FREELIST(ptr) = (struct _Bigint **) _calloc_r (ptr, 
 						      sizeof (struct _Bigint *),
 						      _Kmax + 1);
       if (_REENT_MP_FREELIST(ptr) == NULL)
 	{
+          balloc_unlock();
 	  return NULL;
 	}
     }
@@ -128,6 +133,7 @@
       rv->_k = k;
       rv->_maxwds = x;
     }
+  balloc_unlock();
   rv->_sign = rv->_wds = 0;
   return rv;
 }
@@ -135,12 +141,14 @@
 void
 _DEFUN (Bfree, (ptr, v), struct _reent *ptr _AND _Bigint * v)
 {
+  balloc_lock();
   _REENT_CHECK_MP(ptr);
   if (v)
     {
       v->_next = _REENT_MP_FREELIST(ptr)[v->_k];
       _REENT_MP_FREELIST(ptr)[v->_k] = v;
     }
+  balloc_unlock();
 }
 
 _Bigint *
@@ -432,12 +440,14 @@
   if (!(k >>= 2))
     return b;
   _REENT_CHECK_MP(ptr);
+  balloc_lock();
   if (!(p5 = _REENT_MP_P5S(ptr)))
     {
       /* first time */
       p5 = _REENT_MP_P5S(ptr) = i2b (ptr, 625);
       p5->_next = 0;
     }
+  balloc_unlock();
   for (;;)
     {
       if (k & 1)
